pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java

```diff
@@ -578,21 +582,34 @@ public CompletableFuture<Void> closeAsync() {
         }
     }
 
+    private void close(AutoCloseable closeable, String resource, ScheduledExecutorService shutdownExecutor) {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+        shutdownExecutor.execute(() -> {
+            try {
+                closeable.close();
+                future.complete(null);
+            } catch (Exception e) {
+                LOG.warn("Failed to close {}", resource, e);
+                future.completeExceptionally(new IllegalStateException("Failed to close " + resource));
+            }
+        });
+        CompletableFuture<Void> result = FutureUtil.addTimeoutHandling(future,
+                Duration.ofMillis(Math.max(1L, config.getBrokerShutdownTimeoutMs())), shutdownExecutor,
+                () -> new TimeoutException("Timeout to close " + resource));
+        try {
+            result.join();
```

Comment:
Maybe still exist problem,  this operation is sync wait. Every resource close always wait, the total time over `brokerShutdownTimeoutMs`.